### PR TITLE
[UNR-4232] Subviews remove registered dispatcher callbacks on shutdown

### PR DESCRIPTION
**Pull request by martin-improbable**
#### Description
On shutdown, subviews remove registered dispatcher callbacks.

#### Tests
Automated test.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/SpatialView/ScopedDispatcherCallback.cpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+namespace SpatialGDK
+{
+FScopedDispatcherCallback::FScopedDispatcherCallback(FDispatcher& InDispatcher, const CallbackId InCallbackId)
+	: Dispatcher(&InDispatcher)
+	, ScopedCallbackId(InCallbackId)
+{
+	check(Dispatcher);
+}
+
+FScopedDispatcherCallback::~FScopedDispatcherCallback()
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialView/ScopedDispatcherCallback.cpp:16]: With the way this is setup this type is unsafe to use outside of a unique_ptr, which is likely to make this used incorrectly later.

This should probably be explicitly deleting the copy constructor/assignment and defining a move constructor such that it either nulls out the dispatcher or sets the ID to 0. Then in the destructor we can check either of these conditions and only remove the callback if they are set. 

Alternatively we can add in a a unique_ptr internally although I'm not sure that works without making some heap allocation which seems excessive.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/SpatialView/ScopedDispatcherCallback.cpp:16]: Ah wait, I missed that this is explicitly deleing the move constructors too. Which I guess makes it safe, but also makes it a bigger pain that it needs to be. And means you don't use it as you'd intuitively expect.

Comment (martin-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/SpatialView/ScopedDispatcherCallback.cpp:16]: I was going to add a move constructor if/when we need it, I don't believe I need it at the moment.

Comment (martin-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/SpatialView/ScopedDispatcherCallback.cpp:16]: Happy to add one now if you prefer.

**Review by jacquese — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/SpatialView/ScopedDispatcherCallback.cpp (hunk 1 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/SpatialView/ScopedDispatcherCallback.cpp:16]: I'd prefer to just do it now. I don't see any advantage to not doing it and it's only a few lines of code and it makes the use of the type much more intuitive.

**Review by samiwh — COMMENTED** [review]
Generally lgtm. Comments are only about tests and nits.

File: SpatialGDK/Source/SpatialGDK/Public/SpatialView/DispatcherInterface.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+struct FEntityComponentChange
+{
+	Worker_EntityId EntityId;
+	const ComponentChange& Change;
+};
+
+using FEntityCallback = TCallbacks<Worker_EntityId>::CallbackType;
+using FComponentValueCallback = TCallbacks<FEntityComponentChange>::CallbackType;
+
+class IDispatcher
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialView/DispatcherInterface.h:19]: Nit - I think elsewhere we've used AbstractX (although my preference is always Dispatcher for the base and something more specific for the derived).
But I don't really care that much. Up to you if this is actually important enough to change.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialView/DispatcherInterface.h:19]: Actually if we are doing FStuff etc then I could be convinced we should change other things to IStuff too.

**Review by samiwh — APPROVED** [review]
lgtm

**Pull request merged**